spatial/barneshut/barneshut2.go

```diff
@@ -0,0 +1,249 @@
+	root tile
+
+	Particles []Particle2
+}
+
+// NewPlane returns a new Plane.
+func NewPlane(p []Particle2) *Plane {
+	q := Plane{Particles: p}
+	q.Reset()
+	return &q
+}
+
+// Reset reconstructs the Barnes-Hut tree. Reset must be called if the
+// Particles field or elements of Particles have been altered, unless
+// ForceOn is called with theta=0 or no data structures have been
+// previously built.
+func (q *Plane) Reset() {
+	if len(q.Particles) == 0 {
+		q.root = tile{}
+		return
+	}
+
+	q.root = tile{
+		particle: q.Particles[0],
+		center:   q.Particles[0].Coord2(),
+		mass:     q.Particles[0].Mass(),
+	}
+	q.root.bounds.Min = q.root.center
+	q.root.bounds.Max = q.root.center
+	for _, e := range q.Particles[1:] {
+		c := e.Coord2()
+		if c.X < q.root.bounds.Min.X {
+			q.root.bounds.Min.X = c.X
+		}
+		if c.X > q.root.bounds.Max.X {
+			q.root.bounds.Max.X = c.X
+		}
+		if c.Y < q.root.bounds.Min.Y {
+			q.root.bounds.Min.Y = c.Y
+		}
+		if c.Y > q.root.bounds.Max.Y {
+			q.root.bounds.Max.Y = c.Y
+		}
+	}
+
+	// TODO(kortschak): Partially parallelise this by
+	// choosing the direction and using one of four
+	// goroutines to work on each root quadrant.
+	for _, e := range q.Particles[1:] {
+		q.root.insert(e)
+	}
+	q.root.summarize()
+}
+
+// ForceOn returns a force vector on p given p's mass and the force function, f,
+// using the Barnes-Hut theta approximation parameter.
+//
+// Calls to f will include p in the p1 position and a non-nil p2 if the force
+// interaction is with a non-aggregate mass center, otherwise p2 will be nil.
+//
+// It is safe to call ForceOn concurrently.
+func (q *Plane) ForceOn(p Particle2, theta float64, f Force2) (force r2.Vec) {
+	var empty tile
+	if theta > 0 && q.root != empty {
+		return q.root.forceOn(p, p.Coord2(), p.Mass(), theta, f)
+	}
+
+	// For the degenerate case, just iterate over the
+	// slice of particles rather than walking the tree.
+	var v r2.Vec
+	m := p.Mass()
+	pv := p.Coord2()
+	for _, e := range q.Particles {
+		v = v.Add(f(p, e, m, e.Mass(), e.Coord2().Sub(pv)))
+	}
+	return v
+}
+
+// tile is a quad tree quadrant with Barnes-Hut extensions.
+type tile struct {
+	particle Particle2
+
+	bounds r2.Box
+
+	nodes [4]*tile
+
+	center r2.Vec
+	mass   float64
+}
+
+// insert inserts p into the subtree rooted at t.
+func (t *tile) insert(p Particle2) {
+	if t.particle == nil {
+		for _, q := range t.nodes {
+			if q != nil {
+				t.passDown(p)
+				return
+			}
+		}
+		t.particle = p
+		t.center = p.Coord2()
+		t.mass = p.Mass()
+		return
+	}
+	t.passDown(p)
+	t.passDown(t.particle)
+	t.particle = nil
+	t.center = r2.Vec{}
+	t.mass = 0
+}
+
+func (t *tile) passDown(p Particle2) {
+	dir := quadrantOf(t.bounds, p)
+	if t.nodes[dir] == nil {
+		t.nodes[dir] = &tile{bounds: splitPlane(t.bounds, dir)}
+	}
+	t.nodes[dir].insert(p)
+}
+
+const (
+	ne = iota
+	se
+	sw
+	nw
+)
+
+// quadrantOf returns which quadrant of b that p should be placed in.
+func quadrantOf(b r2.Box, p Particle2) int {
```

Comment:
This could take `p` as `r2.Vec` at the price of less informative panic message.

Reply:
That doesn't really save anything and I think the panic message will likely help users. `passDown` needs to hold a `Particle[23]` in order to insert it, so all we can do is hoist the `Coord[23]` call up one stack level.

Reply:
No problem in leaving as is, I just wanted to read your comment.